src/_posts/2020-05-19-promises-and-lies.md

```diff
@@ -0,0 +1,151 @@
+---
+layout: post
+asset-type: post
+name: promises-and-lies
+title: Promises and Lies
+date: 2020-05-19 10:00:00 +00:00
+author: Christopher Eyre
+description: Adding timeouts to promises in Javascript.
+image:
+    src: /assets/custom/img/blog/lie.png
+    attribution:
+        text: Public domain photo from Pixabay
+        href: https://pixabay.com/photos/lying-promises-deception-dishonesty-1562272/
+tags:
+    - javascript
+    - node
+    - reliability
+abstract: Adding timeouts to promises in Javascript. Introducing lie.js 
+alias: [/2020/05/19/promises-and-lies]
+---
+
+## Promises in Javascript
+
+Promises were added to Javascript as a standard means to avoid nesting callback functions.
+Before promises each function could have it's own functions that would be called once the job is completed. They are mostly used for asynchronous events. They are not a means of lazy evaluation or for process control. Functions that have an async result typically now return a Promise.
+
+Promises have the following format:
+
+```javascript
+var promiseObj = new Promise(executor);
+```
+
+where executor is a function that looks like:
+
+```javascript
+function(resolutionFunc, rejectionFunc){
+    // typically, some asynchronous operation.
+}
+``` 
+
+Promises are used like this:
+
+```javascript
+promiseObj.then(console.log('It worked'))
+          .catch(console.log('It failed'))
+          .finally( console.log('It is over') )
+```
+
+You would typically do more in each block but this gives the idea. It is possible to return a promise from a promise so that they can be chained. Note that the catch at the end of a chain will catch whichever reject happens first.
+
+Once a promise has resolved to a value, then it will always resolve to the same value.
+You can therefore call then on a given promise multiple times and it will always return the same result. I'll come back to this concept later.
+
+Promises also have no means of being cancelled. The only output of them is through actions of the resolve or reject functions. This means that it is possible to cache a promise - just note that the cached promise could have already failed!
+
+## The Promise class
+
+The javascript promise class has the following static functions:
+
+```Javascript
+Promise.all( ) 
+```
+
+This will turn an iteratable of Promises into a Promise that will either return an iteratable of the success values or the first failure condition.
```

Comment:
Spelling:

This will turn iterable of Promises into a Promise that will either return an iterable of the success values or the first failure condition.